sdt_probe_offset.c

```diff
@@ -126,87 +152,314 @@ long get_sdt_probe_offset(int fd, char *probe_name)
 	if (ret) {
 		fprintf(stderr, "ELF get header index failed: %s.\n", elf_errmsg(-1));
 		ret = -1; 
 		goto err2;
 	}
 
 	elf_section = NULL;
-	elf_data = NULL;
+	elf_section_data_desc = NULL;
+	stap_note_section_found = 0;
 
+	/*
+	 * Search ELF sections for stap note section which contains probe
+	 * descriptions.
+	 */
 	while ((elf_section = elf_nextscn(elf_handle, elf_section)) != NULL) {
-		if (gelf_getshdr(elf_section, &elf_section_header) != &elf_section_header) {
-			fprintf(stderr, "GELF get section header failed: %s.\n", elf_errmsg(-1));
+		size_t next_note, note_offset, note_name_offset, note_desc_offset,
+			size_probe_data;
+		char *section_data_ptr, *elf_format;
+		GElf_Nhdr note_hdr;
+
+		/*
+		 * System is assumed to be 64 bit.
+		 * TODO Add support for 32 bit systems
+		 * TODO Change probe_data array to probe_addr and only translate the
+		 * address
+		 */
+		Elf64_Addr probe_data[3];
+
+		/*
+		 * Will contain the in-file and in-memory representations of the probe
+		 * data.
+		 */
+		Elf_Data probe_data_in_file, probe_data_in_mem;
+
+		if (gelf_getshdr(elf_section, &elf_section_hdr) == NULL) {
+			fprintf(stderr,
+				"GELF get section header failed: %s.\n", elf_errmsg(-1));
 			ret = -1;
 			goto err2;
 		}
 
-		if ((section_name = elf_strptr(elf_handle, section_index, elf_section_header.sh_name)) == NULL) {
-			fprintf(stderr, "ELF retrieve string pointer failed: %s.\n", elf_errmsg(-1));
+		section_name = elf_strptr(elf_handle, section_idx, elf_section_hdr.sh_name);
+		if (section_name == NULL) {
+			fprintf(stderr,
+				"ELF retrieve string pointer failed: %s.\n", elf_errmsg(-1));
 			ret = -1;
 			goto err2;
 		}
 
-		if (strcmp(section_name, ".note.stapsdt") != 0) {
+		if (strncmp(section_name, NOTE_STAPSDT_STR, sizeof(NOTE_STAPSDT_STR)) != 0) {
 			continue;
 		}
 
-		elf_data = elf_getdata(elf_section, NULL);
+		stap_note_section_found = 1;
+
+		elf_section_data_desc = elf_getdata(elf_section, NULL);
+		if (elf_section_data_desc == NULL) {
+			fprintf(stderr, "ELF get data failed: %s.\n", elf_errmsg(-1));
+			ret = -1;
+			goto err2;
+		}
+
+		size_probe_data = gelf_fsize(elf_handle, ELF_T_ADDR,
+			ARRAY_LEN(probe_data), EV_CURRENT);
+
+		probe_data_in_mem.d_buf = &probe_data;
+		probe_data_in_mem.d_type = ELF_T_ADDR;
+		probe_data_in_mem.d_version = EV_CURRENT;
+		probe_data_in_mem.d_size = size_probe_data;
 
-		size_t next;
-		GElf_Nhdr nhdr;
-		size_t name_offset;
-		size_t desc_offset;
+		probe_data_in_file.d_buf = NULL;
+		probe_data_in_file.d_type = ELF_T_ADDR;
+		probe_data_in_file.d_version = EV_CURRENT;
+		probe_data_in_file.d_size = probe_data_in_mem.d_size;
 
-		name = "";
+		section_data_ptr = (char*)elf_section_data_desc->d_buf;
+		note_offset = 0;
+		probe_provider_found = 0;
+		probe_name_found = 0;
+		note_probe_provider = "";
+		note_probe_name = "";
+		next_note = gelf_getnote(elf_section_data_desc, note_offset, &note_hdr,
+			&note_name_offset, &note_desc_offset);
```

Comment:
Could you add a few tabs here so that elf_section_data_desc and &note_name_offset line up nicely?